internal/driver/rest/api.go

```diff
@@ -0,0 +1,157 @@
+package rest
+
+import (
+	"encoding/json"
+	"github.com/isdzulqor/donation-hub/internal/core/service/project"
+	"github.com/isdzulqor/donation-hub/internal/core/service/user"
+	"github.com/isdzulqor/donation-hub/internal/driver/rest/req"
+	"github.com/jmoiron/sqlx"
+	"gopkg.in/validator.v2"
+	"log"
+	"net/http"
+	"regexp"
+)
+
+type API struct {
+	DB             *sqlx.DB
+	UserService    user.Service
+	ProjectService project.Service
+}
+
+func (api *API) HandlePostUserRegister(w http.ResponseWriter, r *http.Request) {
+	log.Println("register")
+
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusNotFound)
+		_ = json.NewEncoder(w).Encode(NewNotFound())
+		return
+	}
+
+	var rb req.RegisterReqBody
+	err := json.NewDecoder(r.Body).Decode(&rb)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		_ = json.NewEncoder(w).Encode(NewBadRequest(err.Error()))
+		return
+	}
+
+	// validation
+	isFails := validator.Validate(rb)
+	if isFails != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		_ = json.NewEncoder(w).Encode(NewBadRequest(isFails.Error()))
+		return
+	}
+
+	if !regexp.MustCompile(`^(donor|requester)$`).MatchString(rb.Role) {
+		w.WriteHeader(http.StatusBadRequest)
+		_ = json.NewEncoder(w).Encode(NewBadRequest("role must be 'donor' or 'requester'"))
+		return
+	}
+
+	// store
+	u, err := api.UserService.Register(r.Context(), rb)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		_ = json.NewEncoder(w).Encode(NewBadRequest(err.Error()))
+	}
+
+	_ = json.NewEncoder(w).Encode(u)
+	return
+}
+
+func (api *API) HandlePostUserLogin(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
```

Comment:
This HTTP method checking mechanism should not be done here. It should be easily defined on API list as I mentioned above, also apply to all handlers